fiftyone/utils/youtube.py

```diff
@@ -0,0 +1,461 @@
+    specify a deisred video codec and resolution to download, if possible.
+
+    YouTube videos are regularly taken down. Therefore, this method provides an
+    optional ``max_videos`` argument that you can use in conjunction with
+    ``skip_failures=True`` and a large list of possibly non-existent videos in
+    ``urls`` in cases where you need a certain number of videos to be
+    successfully downloaded but are willing to tolerate failures.
+
+    Args:
+        urls: a list of YouTube URLs to download
+        download_dir (None): a directory in which to store the downloaded
+            videos
+        video_paths (None): a list of paths to which to download the videos.
+            When downloading entire videos, a stream matching the video format
+            implied by each file's extension is downloaded, if available, or
+            else the extension of the video path is **changed** to match the
+            available stream's format
+        clip_segments (None): a list of ``(first, last)`` tuples defining a
+            specific segment of each video to download
+        ext (".mp4"): a video format to download for each video, if possible.
+            Only applicable when a ``download_dir`` is used. This format will
+            be respected if such a stream exists, otherwise the format of the
+            best available stream is used. Set this value to ``None`` if you
+            want to download the stream with the best match for ``resolution``
+            and ``progressive`` regardless of format
+        only_progressive (True): whether to only download progressive streams,
+            if possible. Progressive streams contain both audio and video
+            tracks and are typically only available at <= 720p resolution
+        resolution ("highest"): a desired stream resolution to download. This
+            filter is applied after respecting the desired video format and
+            ``only_progressive`` restriction, if applicable. The supported
+            values are:
+
+            -   ``"highest"`` (default): download the highest resolution stream
+            -   ``"lowest"``:  download the lowest resolution stream
+            -   A target resolution like ``"1080p"``. In this case, the stream
+                whose resolution is closest to this target value is downloaded
+        max_videos (None): the maximum number of videos to successfully
+            download. By default, all videos are be downloaded
+        num_workers (None): the number of threads or processes to use when
+            downloading videos. By default, ``multiprocessing.cpu_count()`` is
+            used
+        skip_failures (True): whether to gracefully continue without raising
+            an error if a video cannot be downloaded
+        quiet (False): whether to suppress logging, except for a progress bar
+
+    Returns:
+        a tuple of
+
+        -   **downloaded**: a dict mapping integer indexes into ``urls`` to
```

Comment:
I think a return dict structure like this is, in general, the only way to succinctly describe what was downloaded and what failed in a situation where `len(urls) > max_videos` and certain videos are not available in the requested `ext` and thus the user needs a way to deduce the names of the returned videos.